foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/expressions/SQLSelectStatement.java

```diff
@@ -1722,69 +1722,102 @@ protected void normalizeOrderBy(Expression builder, List<Expression> allExpressi
-                    printer.printString("(");
-                }
-            }
-            printer.printString("SELECT ");
+            Vector selectFields = null;
+            selectFields = printSQLSelect(printer);
+            printSQLWhereKeyWord(printer);
+            printSQLWhereClause(printer);
+            printSQLHierarchicalQueryClause(printer);
+            printSQLGroupByClause(printer);
+            printSQLHavingClause(printer);
+            printSQLOrderByClause(printer);
+            printSQLForUpdateClause(printer);
+            printSQLUnionClause(printer);
+            return selectFields;
+        } catch (IOException exception) {
+            throw ValidationException.fileError(exception);
+        }
+    }
 
-            if (getHintString() != null) {
-                printer.printString(getHintString());
-                printer.printString(" ");
-            }
+    public Vector printSQLSelect(ExpressionSQLPrinter printer) throws IOException {
+        Vector selectFields = null;
+        printer.setRequiresDistinct(shouldDistinctBeUsed());
 
-            if (shouldDistinctBeUsed()) {
-                printer.printString("DISTINCT ");
+        if (hasUnionExpressions()) {
+            // Ensure union order using brackets.
+            int size = getUnionExpressions().size();
+            for (int index = 0; index < size; index++) {
+                printer.printString("(");
             }
+        }
+        printer.printString("SELECT ");
 
-            selectFields = writeFieldsIn(printer);
-            //fix bug:6070214: turn off unique field aliases after fields are written
-            setUseUniqueFieldAliases(false);
+        if (getHintString() != null) {
+            printer.printString(getHintString());
+            printer.printString(" ");
+        }
 
-            appendFromClauseToWriter(printer);
+        if (shouldDistinctBeUsed()) {
+            printer.printString("DISTINCT ");
+        }
 
-            if (!(getWhereClause() == null)) {
-                printer.printString(" WHERE ");
-                printer.printExpression(getWhereClause());
-            }
+        selectFields = writeFieldsIn(printer);
+        //fix bug:6070214: turn off unique field aliases after fields are written
+        setUseUniqueFieldAliases(false);
 
-            if (hasHierarchicalQueryExpressions()) {
-                appendHierarchicalQueryClauseToWriter(printer);
-            }
+        appendFromClauseToWriter(printer);
+        return selectFields;
+    }
 
-            if (hasGroupByExpressions()) {
-                appendGroupByClauseToWriter(printer);
-            }
-            if (hasHavingExpression()) {
-                //appendHavingClauseToWriter(printer);
-                printer.printString(" HAVING ");
-                printer.printExpression(getHavingExpression());
-            }
+    public void printSQLWhereKeyWord(ExpressionSQLPrinter printer) throws IOException {
```

Comment:
do these new methods have to be public instead of private/protected?

Reply:
Yes because they are called from `org.eclipse.persistence.platform.database.OraclePlatform` (different package).

Reply:
ok